src/Shared/InterningBinaryReader.cs

```diff
@@ -146,15 +150,24 @@ override public String ReadString()
                         charsRead = _decoder.GetChars(rawBuffer, rawPosition, n, charBuffer, 0);
                         return Strings.WeakIntern(charBuffer.AsSpan(0, charsRead));
                     }
-
+#if !NET35
+                    resultBuffer ??= ArrayPool<char>.Shared.Rent(stringLength); // Actual string length in chars may be smaller.
+#else
+                    // Since NET35 is only used in rare TaskHost processes, we decided to leave it as-is.
                     resultBuffer ??= new char[stringLength]; // Actual string length in chars may be smaller.
+#endif
                     charsRead += _decoder.GetChars(rawBuffer, rawPosition, n, resultBuffer, charsRead);
 
                     currPos += n;
                 }
                 while (currPos < stringLength);
 
-                return Strings.WeakIntern(resultBuffer.AsSpan(0, charsRead));
+                var retval = Strings.WeakIntern(resultBuffer.AsSpan(0, charsRead));
+#if !NET35
+                // resultBuffer is never null, so we can simply return it to the Pool
+                ArrayPool<char>.Shared.Return(resultBuffer);
```

Comment:
nit: Would it be safer to do this inside a `finally`?

Reply:
I have thought about it, but decided to prioritize code simplicity and readability here. The consequence of exceptions being thrown are buffer not being release are small. ArrayPool is not holding reference to Rented buffers. If buffer is not Returned it will be collected by GC results to equivalent of `new T[]`. So in worst case it will regress to previous behavior i.e. buffer allocation per call.

Reply:
On the other hand. `finally` version seems to be safer for code maintenance...

Changed